### DISCUSSION TITLE
rewriting-git-history-simply-with-git-filter-repo/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#122] [Announcements]
# rewriting-git-history-simply-with-git-filter-repo/

In this post I describe how I used git-filter-repo in Docker to rewrite the history of a git repository to move files into a subfolder

https://andrewlock.net/rewriting-git-history-simply-with-git-filter-repo/

**Comment by kkm000** [comment]
Hi Andrew, thank you for suggesting this powerful tool!

> As long as you have your work pushed to a remote...

I wanted to add that you do not need in fact push the repo to a remote to obtain a fresh clone. Git clones repos from the filesystem just fine. What I in fact did, was

```

mkdir ~/rewrite  # Temporary local clone directory
cd ~/rewrite
git clone --no-hardlinks ~/work .  # A path to repo to clone.
  Cloning into '.'...
  done.
git remote remove origin  # ~/work had been recorded as a remote when cloning.
```
And here's my fresh clone! After everything is confirmed fine, and assuming the original (`~/work`) worktree had everything checked in, had no unchecked imported files and no remote tracking branches (or you're fine with re-creating them; you're up to a force push anyway), it's ok to remove it and simply move the temporary `rewrite` clone to its place, only preserve its local `config`:

```

cp ~/work/.git/config ~/rewrite/.git/
rm -rf ~/work
mv ~/rewrite ~/work
```

**Comment by charlesr1971** [comment]
Thanks for this great article. I am not sure why you are having a problem installing Python to Windows. I have Python on Windows 10, 11 & Server 2019? 
And the installation  was completely painless.

**Reply by andrewlock** [reply]
Python 2 Vs 3 is where the pain starts in my experience 😉 admittedly that becomes less of an issue as time goes on!

**Comment by sarnobat** [comment]
After a bit of trial and improvement, I found that my 15G non-bare repo could be reduced to 8.4G non-bare if I removed all files bigger than 5M.

The command I used was:
```
git filter-repo --force --invert-paths --paths-from-file /tmp/git_remove_5M_plus.txt
```
Note that paths in the txt file should be relative, and NOT contain the prefix `./` (so you need to remove those from the output of `find -size +5`)
